ui/src/app/store/zone/slice.ts

```diff
@@ -1,22 +1,322 @@
+import type { PayloadAction } from "@reduxjs/toolkit";
 import { createSlice } from "@reduxjs/toolkit";
 
-import { ZoneMeta } from "./types";
-import type { CreateParams, UpdateParams, ZoneState } from "./types";
-
 import {
-  generateCommonReducers,
-  genericInitialState,
-} from "app/store/utils/slice";
+  ACTION_STATUS,
+  ZONE_ACTIONS,
+  ZONE_MODEL,
+  ZONE_PK,
+  ZONE_WEBSOCKET_METHODS,
+} from "./constants";
+import type {
+  ActionStatuses,
+  CreateParams,
+  DeleteParams,
+  UpdateParams,
+  Zone,
+  ZoneActionNames,
+  ZoneGenericActions,
+  ZoneModelActions,
+  ZonePK,
+  ZoneState,
+} from "./types";
+
+import type { APIError } from "app/base/types";
+
+type ActionWithMeta<P = null> = PayloadAction<P, string, { modelPK: ZonePK }>;
+
+const {
+  cleanup: cleanupAction,
+  create: createAction,
+  delete: deleteAction,
+  fetch: fetchAction,
+  update: updateAction,
+} = ZONE_ACTIONS;
+
+const {
+  failed: actionFailed,
+  processing: actionProcessing,
+  successful: actionSuccessful,
+} = ACTION_STATUS;
+
+export const initialGenericActions: ZoneGenericActions = {
+  [createAction]: {
+    [actionFailed]: false,
+    [actionProcessing]: false,
+    [actionSuccessful]: false,
+  },
+  [fetchAction]: {
+    [actionFailed]: false,
+    [actionProcessing]: false,
+    [actionSuccessful]: false,
+  },
+};
+
+export const initialModelActions: ZoneModelActions = {
+  [deleteAction]: {
+    [actionFailed]: [],
+    [actionProcessing]: [],
+    [actionSuccessful]: [],
+  },
+  [updateAction]: {
+    [actionFailed]: [],
+    [actionProcessing]: [],
+    [actionSuccessful]: [],
+  },
+};
+
+const addError = (
+  state: ZoneState,
+  action: ZoneActionNames,
+  error: APIError,
+  modelPK: ZonePK | null = null
+) => {
+  state.errors.unshift({ action, error, modelPK });
+};
+
+const updateGenericAction = (
+  state: ZoneState,
+  actionName: keyof ZoneGenericActions,
+  status: ActionStatuses
+) => {
+  const statuses = state.genericActions[actionName];
+  if (status === actionFailed) {
+    statuses[actionFailed] = true;
+    statuses[actionProcessing] = false;
+    statuses[actionSuccessful] = false;
+  } else if (status === actionProcessing) {
+    statuses[actionFailed] = false;
+    statuses[actionProcessing] = true;
+    statuses[actionSuccessful] = false;
+  } else {
+    statuses[actionFailed] = false;
+    statuses[actionProcessing] = false;
+    statuses[actionSuccessful] = true;
+  }
```

Comment:
Looks like only one of the statuses can be true at any given moment. Why not return a single status string?

Reply:
+1